### PR TITLE
Filter blocks to add to redstone sim's wake queue

### LINKED ISSUE #4402: Crash on loading world
Client version:   1.12.2
Server OS:        Windows

### Actual behavior
Server crashed on loading world. Tried many times to reproduce it no luck.

```
>	[Inline Frame] Cuberite.exe!std::_Uninitialized_move(Vector3<int> * const) Line 1697	C++
 	[Inline Frame] Cuberite.exe!std::vector<Vector3<int>,std::allocator<Vector3<int> > >::_Umove(Vector3<int> *) Line 1558	C++
 	Cuberite.exe!std::vector<Vector3<int>,std::allocator<Vector3<int> > >::_Emplace_reallocate<Vector3<int> const &>(Vector3<int> * const _Whereptr, const Vector3<int> & <_Val_0>) Line 708	C++
 	[Inline Frame] Cuberite.exe!std::vector<Vector3<int>,std::allocator<Vector3<int> > >::emplace_back(const Vector3<int> &) Line 661	C++
 	[Inline Frame] Cuberite.exe!std::vector<Vector3<int>,std::allocator<Vector3<int> > >::push_back(const Vector3<int> &) Line 670	C++
 	[Inline Frame] Cuberite.exe!cIncrementalRedstoneSimulatorChunkData::WakeUp(const Vector3<int> &) Line 18	C++
 	Cuberite.exe!cIncrementalRedstoneSimulator::AddBlock(Vector3<int> a_Block, cChunk * a_Chunk) Line 37	C++
 	Cuberite.exe!cSimulator::WakeUp(Vector3<int> a_Block, cChunk * a_Chunk) Line 36	C++
 	[Inline Frame] Cuberite.exe!cChestEntity::OnSlotChanged::__l5::<lambda>(cChunk &) Line 96	C++
 	Cuberite.exe!cFunctionRef<bool __cdecl(cChunk &)>::ObjectFunctionCaller<bool <lambda>(cChunk &) >(void * a_Callable, cChunk & <a_Args_0>) Line 71	C++
 	[Inline Frame] Cuberite.exe!cFunctionRef<bool __cdecl(cChunk &)>::operator()(cChunk &) Line 57	C++
 	Cuberite.exe!cChunkMap::DoWithChunk(int a_ChunkX, int a_ChunkZ, cFunctionRef<bool __cdecl(cChunk &)> a_Callback) Line 305	C++
 	[Inline Frame] Cuberite.exe!cChunkMap::DoWithChunkAt(Vector3<int>) Line 316	C++
 	[Inline Frame] Cuberite.exe!cWorld::DoWithChunkAt(Vector3<int>) Line 1568	C++
 	Cuberite.exe!cChestEntity::OnSlotChanged(cItemGrid * a_Grid, int a_SlotNum) Line 101	C++
 	Cuberite.exe!cItemGrid::TriggerListeners(int a_SlotNum) Line 824	C++
 	Cuberite.exe!cItemGrid::SetSlot(int a_SlotNum, const cItem & a_Item) Line 153	C++
 	Cuberite.exe!cWSSAnvil::LoadItemGridFromNBT(cItemGrid & a_ItemGrid, const cParsedNBT & a_NBT, int a_ItemsTagIdx, int) Line 869	C++
 	Cuberite.exe!cWSSAnvil::LoadChestFromNBT(const cParsedNBT & a_NBT, int a_TagIdx, unsigned char a_BlockType, unsigned char a_BlockMeta, int a_BlockX, int a_BlockY, int a_BlockZ) Line 1095	C++
 	Cuberite.exe!cWSSAnvil::LoadBlockEntityFromNBT(const cParsedNBT & a_NBT, int a_Tag, int a_BlockX, int a_BlockY, int a_BlockZ, unsigned char a_BlockType, unsigned char a_BlockMeta) Line 738	C++
 	Cuberite.exe!cWSSAnvil::LoadBlockEntitiesFromNBT(std::map<int,cBlockEntity *,std::less<int>,std::allocator<std::pair<int const ,cBlockEntity *> > > & a_BlockEntities, const cParsedNBT & a_NBT, int a_TagIdx, unsigned char * a_BlockTypes, unsigned char * a_BlockMetas) Line 672	C++
 	Cuberite.exe!cWSSAnvil::LoadChunkFromNBT(const cChunkCoords & a_Chunk, const cParsedNBT & a_NBT, const std::basic_string<char,std::char_traits<char>,std::allocator<char> > & a_RawChunkData) Line 424	C++
 	Cuberite.exe!cWSSAnvil::LoadChunkFromData(const cChunkCoords & a_Chunk, const std::basic_string<char,std::char_traits<char>,std::allocator<char> > & a_Data) Line 333	C++
 	Cuberite.exe!cWSSAnvil::LoadChunk(const cChunkCoords & a_Chunk) Line 149	C++
 	Cuberite.exe!cWorldStorage::LoadChunk(int a_ChunkX, int a_ChunkZ) Line 293	C++
 	[Inline Frame] Cuberite.exe!cWorldStorage::LoadOneChunk() Line 238	C++
 	Cuberite.exe!cWorldStorage::Execute() Line 217	C++
 	[External Code]	
 	[Inline Frame] Cuberite.exe!invoke_thread_procedure(unsigned int(*)(void *)) Line 91	C++
 	Cuberite.exe!thread_start<unsigned int (__cdecl*)(void * __ptr64)>(void * const parameter) Line 115	C++
```

Looking at the stack trace:
* Server loads a chunk containing a chest
* Chest contains items and items are loaded
* Loading the items triggers the `cChestEntity::OnSlotChanged`
* This wakes up the redstone simulator
* A Vector3<int> is pushed into a vector map

Something here is in a invalid state and this crashes the server.

### PR DESCRIPTION
**Pull request by peterbell10**
Fixes #4402

Instead of adding every block in the world to the update queue, this only adds blocks with a redstone mechanism nearby. Or, if we can't read its neighbours I add it anyway just in case.

This still allows for indirect powering through solid blocks but the update queue is a lot smaller in the common case.

### CODE REVIEW

**Comment by madmaxoft** [comment]
Although I normally don't read through the redstone jungle, this is pretty straightforward so even I can review it :)

File: src/Simulator/IncrementalRedstoneSimulator/IncrementalRedstoneSimulator.cpp [hunk 1]
```diff
@@ -193,3 +192,63 @@ void cIncrementalRedstoneSimulator::Simulate(float a_dt)
...
+	{
+		m_Data.WakeUp(a_Block);
+		return;
+	}
+
+	const auto RelPos = cChunkDef::AbsoluteToRelative(a_Block, a_Chunk->GetPos());
+	const auto CurBlock = a_Chunk->GetBlock(RelPos);
+
+	// Always update mechanisms
+	if (IsMechanism(CurBlock))
```
Comment (tigerw) [inline src/Simulator/IncrementalRedstoneSimulator/IncrementalRedstoneSimulator.cpp:213]: Perhaps IsRedstone is better? E.g. it seems like pressure plates will never get updated as they aren't mechanisms.

**Comment by tigerw** [comment]
i yes these changes

**Comment by bearbin** [comment]
On Sun, 05 Apr 2020 09:59:42 -0700
Tiger Wang <notifications@github.com> wrote:

> i yes these changes
> 

But no nice green tick?

**Comment by tigerw** [comment]
my view of GitHub is still stuck back when I last contributed, before all of the reviewers and approvals features got added XD

**Pull request merged**